server.js

```diff
@@ -374,22 +389,46 @@ connection.onDidChangeConfiguration(({ settings }) => {
 	validateLanguages = settings.stylelint.validate || [];
 	snippetLanguages = settings.stylelint.snippet || ['css', 'less', 'postcss', 'scss'];
 
-	const removeLanguages = oldValidateLanguages.filter((lang) => !validateLanguages.includes(lang));
+	const validateLanguageSet = new Set(validateLanguages);
+	const oldValidateLanguageSet = new Set(oldValidateLanguages);
 
-	for (const document of documents
-		.all()
-		.filter((doc) => removeLanguages.includes(doc.languageId))) {
-		clearDiagnostics(document);
+	/** Whether or not the list of languages that should be validated has changed. */
+	let changed = validateLanguageSet.size !== oldValidateLanguageSet.size;
+
+	/** The languages removed from the list of languages that should be validated */
+	const removeLanguages = new Set();
+
+	// Check if the sets are unequal, which means that the list of languages that should be
+	// validated has changed.
+	for (const language of oldValidateLanguageSet) {
+		if (!validateLanguageSet.has(language)) {
+			removeLanguages.add(language);
+			changed = true;
+		}
 	}
 
-	if (removeLanguages.length > 0) {
-		connection.sendNotification('stylelint/languageIdsRemoved', { langIds: removeLanguages });
+	// If languages have been removed, clear diagnostics for documents of those languages.
+	if (removeLanguages.size > 0) {
+		for (const document of documents.all().filter((doc) => removeLanguages.has(doc.languageId))) {
+			clearDiagnostics(document);
+		}
 	}
 
-	const addLanguages = validateLanguages.filter((lang) => !oldValidateLanguages.includes(lang));
+	// If dynamic registration is supported and the list of languages that should be validated
+	// has changed, then (re-)register the formatter.
+	if (registerFormatterDynamically && changed) {
+		// Dispose the old formatter registration if it exists.
+		if (formatterRegistration) {
+			void formatterRegistration.then((disposable) => disposable.dispose());
+		}
 
-	if (addLanguages.length > 0) {
-		connection.sendNotification('stylelint/languageIdsAdded', { langIds: addLanguages });
+		// If there are languages that should be validated, register a formatter for those
+		// languages.
+		if (validateLanguages.length > 0) {
+			formatterRegistration = connection.client.register(DocumentFormattingRequest.type, {
+				documentSelector: validateLanguages,
+			});
```

Comment:
The param to the `register` call must be `DocumentFormattingRegistrationOptions` which is not what is passed here. In other words, `documentSelector` must be a `DocumentFilter[]`, not a `string[]`.

Reply:
Hmm, it could be that the language server library supports more options and just converts the arguments to something expected by the LSP clients. I've only looked at the current LSP spec (https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#documentSelector) which lists those interfaces:

```ts
export interface DocumentFormattingRegistrationOptions extends
	TextDocumentRegistrationOptions, DocumentFormattingOptions {
}

/**
 * General text document registration options.
 */
export interface TextDocumentRegistrationOptions {
	/**
	 * A document selector to identify the scope of the registration. If set to
	 * null the document selector provided on the client side will be used.
	 */
	documentSelector: DocumentSelector | null;
}

export type DocumentSelector = DocumentFilter[];
```

But besides checking the spec, I've also tried this change in a conforming client (ST LSP) and got an error due to wrong argument to registerCapability request.

I'll look closer into it and report back.

Reply:
Thanks for looking into it! It might be a VS Code sugar thing. I think it's also important to note that the types the extension is using [are out of date](https://github.com/stylelint/vscode-stylelint/blob/0ca1019/package.json#L208-L211), which might be another reason for this discrepancy. Either way, the fix is simple so I'm happy to make it if it needs doing.

Reply:
Yeah, I believe the current code is actually wrong. The second argument to `register` will accept anything because the types include `any` but that ends up being [sent verbatim](https://github.com/microsoft/vscode-languageserver-node/blob/10b56de150ad67c3c330da8e2df53ebf2cf347c4/server/src/common/server.ts#L794-L807) through the `client/registerCapablity` request and that is not valid registration request.

Reply:
> Yeah, I believe the current code is actually wrong. The second argument to `register` will accept anything because the types include `any` but that ends up being [sent verbatim](https://github.com/microsoft/vscode-languageserver-node/blob/10b56de150ad67c3c330da8e2df53ebf2cf347c4/server/src/common/server.ts#L794-L807) through the `client/registerCapablity` request and that is not valid registration request.

Just a note that it does check types for the second argument of `register`:

![screenshot](https://user-images.githubusercontent.com/9868643/135907472-92b7e63e-9a6f-4cbc-b473-4d59be2d5fb7.png)

But either way, if this is causing errors in non-VS Code contexts, I'll update the code anyway. I'll push up a commit shortly.

Reply:
Yes, It looks like in the used version of `vscode-languageserver-protocol` there is this definition that allows plain string:

```
export declare type DocumentSelector = (string | DocumentFilter)[];
```

but I believe that variant never made it to the spec so it only works with VSCode.

@dbaeumer can probably confirm that.